syz-manager/rpc.go

```diff
@@ -116,109 +110,29 @@ func (serv *RPCServer) Connect(a *rpctype.ConnectArgs, r *rpctype.ConnectRes) er
 	r.CoverFilterBitmap = createCoverageBitmap(serv.cfg.SysTarget, instCoverFilter)
 	r.EnabledCalls = serv.cfg.Syscalls
-	r.NoMutateCalls = serv.cfg.NoMutateCalls
 	r.GitRevision = prog.GitRevision
 	r.TargetRevision = serv.cfg.Target.Revision
-	if serv.mgr.rotateCorpus() && serv.rnd.Intn(5) == 0 {
-		// We do rotation every other time because there are no objective
-		// proofs regarding its efficiency either way.
-		// Also, rotation gives significantly skewed syscall selection
-		// (run prog.TestRotationCoverage), it may or may not be OK.
-		r.CheckResult = serv.rotateCorpus(f, corpus)
-	} else {
-		r.CheckResult = serv.checkResult
-		f.inputs = corpus
-		f.newMaxSignal = serv.maxSignal.Copy()
-	}
-	return nil
-}
+	r.CheckResult = serv.checkResult
 
-func (serv *RPCServer) rotateCorpus(f *Fuzzer, corpus []rpctype.Input) *rpctype.CheckArgs {
-	// Fuzzing tends to stuck in some local optimum and then it fails to cover
-	// other state space points since code coverage is only a very approximate
-	// measure of logic coverage. To overcome this we introduce some variation
-	// into the process which should cause steady corpus rotation over time
-	// (the same coverage is achieved in different ways).
-	//
-	// First, we select a subset of all syscalls for each VM run (result.EnabledCalls).
-	// This serves 2 goals: (1) target fuzzer at a particular area of state space,
-	// (2) disable syscalls that cause frequent crashes at least in some runs
-	// to allow it to do actual fuzzing.
-	//
-	// Then, we remove programs that contain disabled syscalls from corpus
-	// that will be sent to the VM (f.inputs). We also remove 10% of remaining
-	// programs at random to allow to rediscover different variations of these programs.
-	//
-	// Then, we drop signal provided by the removed programs and also 10%
-	// of the remaining signal at random (f.newMaxSignal). This again allows
-	// rediscovery of this signal by different programs.
-	//
-	// Finally, we adjust criteria for accepting new programs from this VM (f.rotatedSignal).
-	// This allows to accept rediscovered varied programs even if they don't
-	// increase overall coverage. As the result we have multiple programs
-	// providing the same duplicate coverage, these are removed during periodic
-	// corpus minimization process. The minimization process is specifically
-	// non-deterministic to allow the corpus rotation.
-	//
-	// Note: at no point we drop anything globally and permanently.
-	// Everything we remove during this process is temporal and specific to a single VM.
-	calls := serv.rotator.Select()
-
-	var callIDs []int
-	callNames := make(map[string]bool)
-	for call := range calls {
-		callNames[call.Name] = true
-		callIDs = append(callIDs, call.ID)
+	if fuzzer := serv.mgr.getFuzzer(); fuzzer != nil {
+		// A Fuzzer object is created after the first Check() call.
+		// If there was none, there would be no collected max signal either.
+		runner.newMaxSignal = fuzzer.Cover.CopyMaxSignal()
```

Comment:
> I think it's better to return maxCover from fuzzerConnect method (we already return a bunch of info from it to return in the response).

It must definitely happen under `serv.mu.Lock()` -- otherwise we risk races with `distributeSignalDelta()` if the new max signal is distributed before we've saved the new `Runner` object.

Right now the intersection of `serv.mu` and `mgr.mu` is small - only for the time of `getFuzzer()` (one operation). We can even get rid of `serv.mu` here if we use `atomic.Pointer[fuzzer.Fuzzer]`.

If we move `CopyMaxSignal()` to `fuzzerConnect()`, all of `fuzzerConnect()` would have to be both under `serv.mu` and `mgr.mu`. Do we want that?

> And otherwise assume the fuzzer exists for all other getFuzzer() calls.

In `shutdownInstance()` we are safe to ignore the check (as it already is) -- if there was no object in the first place, there would be no `runner.requests` either. 

I'd still keep the check in `ExchangeInfo()` -- normally `syz-fuzzer` waits until `Manager.Check` is finished, so the fuzzer should be in place, but do we want to risk crashing one component (`syz-manager`) in case of a misbehavior of another one? I think one simple `if` check is not a very high price to pay for the increased robustness :)

Reply:
Spent a bit thinking about this.
I think it's most important to keep ExchangeInfo fast and not block/wait for parallel Connect/Shutdown methods. Currently exchange locks serv.mu read runners map, so it can wait for some slower operations on other instances.
If we use sync.Map for runners, then we can structure ExchangeInfo as:

```
func (serv *RPCServer) ExchangeInfo(a *rpctype.ExchangeInfoRequest, r *rpctype.ExchangeInfoReply) error {
	runner, _ := serv.runners.Load(a.Name)
	if runner == nil {
		return nil
	}
	runner.mu.Lock()
	defer runner.mu.Unlock()
	...
}
```

It only holds runner.mu, which is not locked by slower operations on other instances. And then Connect/Shutdown can use simpler, coarser-gained locking.

Reply:
I've switched to `sync.Map` and added a panic in `ExchangeInfo()`.